models/intermediate/int_zendesk__schedule_spine.sql

```diff
@@ -90,21 +95,224 @@ with timezone as (
+        null as holiday_week_end,
+        null as holiday_name_check
     from calculate_schedules
+
+-- Within the spine, determine which schedules contain a holiday
+), valid_adjustment as (
+
+    select 
+        *,
+        case when max(holiday_name_check) over (partition by schedule_id, holiday_week_start order by holiday_week_start rows unbounded preceding) is not null then 1 else 0 end as is_holiday_week
+    from spine_union
+
+-- Now that we have an understanding of which weeks are holiday's let's consolidate them with non holiday weeks
+), all_periods as (
+
+    select distinct
+        schedule_id,
+        holiday_week_start as period_start,
+        holiday_week_end as period_end,
+        start_time_utc,
+        end_time_utc,
+        holiday_name_check,
+        true as is_holiday_week
+    from valid_adjustment
+    where holiday_week_start is not null
+      and holiday_week_end is not null
+
+    union all
+
+    select distinct
+        schedule_id,
+        valid_from as period_start,
+        valid_until as period_end,
+        start_time_utc,
+        end_time_utc,
+        cast(null as {{ dbt.type_string() }}) as holiday_name_check,
+        false as is_holiday_week
+    from valid_adjustment
+
+-- We have holiday and non holiday schedules together, now let's sort them to understand the previous end and next start of neighboring schedules
+), sorted_periods as (
+
+    select distinct
+        *,
+        lag(period_end) over (partition by schedule_id order by period_start, start_time_utc) as prev_end,
+        lead(period_start) over (partition by schedule_id order by period_start, start_time_utc) as next_start
+    from all_periods
+
+-- We need to adjust some non holiday schedules in order to properly fill holiday gaps in the schedules later down the transformation
+), non_holiday_period_adjustments as (
+
+    select
+        schedule_id, 
+        period_start, 
+        period_end,
+        coalesce(case 
+          when not is_holiday_week and prev_end is not null then first_value(prev_end) over (partition by schedule_id, period_start order by period_start, start_time_utc rows unbounded preceding)
+          else period_start
+        end, period_start) as valid_from,
+          coalesce(case 
+          when not is_holiday_week and next_start is not null then last_value(next_start) over (partition by schedule_id, period_start order by period_start, start_time_utc rows unbounded preceding)
+          else period_end
+        end, period_end) as valid_until,
+        start_time_utc,
+        end_time_utc,
+        holiday_name_check,
+        is_holiday_week
+    from sorted_periods
+
+-- There may be gaps in holiday and non holiday schedules, so we need to identify where these gaps are
+), gap_adjustments as(
+
+    select 
+        *,
+        case when valid_from > cast({{ dbt.dateadd("hour", "2", "lag(valid_until) over (partition by schedule_id order by valid_until)") }} as {{ dbt.type_timestamp() }}) 
```

Comment:
could you explain this line a little?

Reply:
Sure thing! Essentially I want to identify where there are gaps in the schedule spine. This code is essentially checking the current record `valid_from` and seeing if it is greater than the previous `valid_until` plus 2 hours. Since the valid_from and valid_until timestamps should match up with each other to the within at least a minute of margin we should not see that the current valid_from is greater than the previous valid_until plus two hours. 

The plus two hours is just a small window to reach forward and see where the two timestamps are in relation to one another. If the valid_from is in fact greater, then that means there is a significant gap in the date_spine. This is how we are able to identify the gaps. 

I am definitely open to other ideas. However, this was the best approach I could think through.

Reply:
Added comment in line.